src/main/java/org/terasology/launcher/game/GameManager.java

```diff
@@ -0,0 +1,150 @@
+        if (Files.notExists(cachedZip)) {
+            download(release, cachedZip, listener);
+        }
+
+        if (!listener.isCancelled()) {
+            final Path extractDir = getInstallDirectory(release.getId());
+            FileUtils.extractZipTo(cachedZip, extractDir);
+            Platform.runLater(() -> installedGames.add(release.getId()));
+            logger.info("Finished installing package: {}", release.getId());
+        }
+    }
+
+    private void download(GameRelease release, Path targetLocation, ProgressListener listener) throws DownloadException, IOException {
+        final URL downloadUrl = release.getUrl();
+
+        final long contentLength = DownloadUtils.getContentLength(downloadUrl);
+        final long availableSpace = targetLocation.getParent().toFile().getUsableSpace();
+
+        if (availableSpace >= contentLength) {
+            final Path cacheZipPart = targetLocation.resolveSibling(targetLocation.getFileName().toString() + ".part");
+            Files.deleteIfExists(cacheZipPart);
+            DownloadUtils.downloadToFile(downloadUrl, cacheZipPart, listener);
+
+            if (!listener.isCancelled()) {
+                Files.move(cacheZipPart, targetLocation, StandardCopyOption.ATOMIC_MOVE);
+            }
+        } else {
+            throw new DownloadException("Insufficient space for downloading package");
+        }
+
+        logger.info("Finished downloading package: {}", release.getId());
+    }
+
+    /**
+     * Removes the given release from the local file system.
+     *
+     * @param game the game release to be removed
+     */
+    public void remove(GameIdentifier game) throws IOException {
+        // Recursively delete all files
+        Files.walk(getInstallDirectory(game))
+                .sorted(Comparator.reverseOrder())
+                .map(Path::toFile)
+                .forEach(File::delete);
+
+        Platform.runLater(() -> installedGames.remove(game));
+        logger.info("Finished removing package: {}", game);
+    }
+
+    /**
+     * Retrieve the set of installed games as observable set.
+     *
+     * @return the set of all installed game releases
+     */
+    public ObservableSet<GameIdentifier> getInstalledGames() {
+        return installedGames;
+    }
+
+    public Path getInstallDirectory(GameIdentifier id) {
+        return installDirectory.resolve(id.getProfile().name()).resolve(id.getBuild().name()).resolve(id.getVersion());
+    }
+
+    /**
+     * Scans the installation directory and collects the installed games.
+     */
+    private void scanInstallationDir() {
+        Set<GameIdentifier> localGames = new HashSet<>();
+        if (Files.exists(installDirectory)) {
+            for (File profileDirectory : Objects.requireNonNull(installDirectory.toFile().listFiles())) {
+                Profile profile;
+                try {
+                    profile = Profile.valueOf(profileDirectory.getName());
```

Comment:
Should we maybe rather have a profile enum and only scan directories that match one of the values in that enum? otherwise we might end up adding a game identifier with profile "foo". On the other hand I'm not sure if that would be a problem :thinking:

Reply:
Oh, `Profile.valueOf(...)` will throw an exception if the directory name is not a valid member of the `Profile` enum. So currently we scan _all_ directories, but only continue processing the "right" ones (the ones we understand). 

Your suggested approach should work equally well. Both have pros and cons (yours probably has better performance, and I was thinking about adding better logging or even migration handling here to recover some old installations).

Reply:
Wasn't aware that `Profile.valueOf(...)` does that, but in that case all is well :+1: